vsp.go

```diff
@@ -57,65 +61,325 @@ func (mw *MultiWallet) NewVSPClient(vspHost string, walletID int, purchaseAccoun
+				return nil
+			}
+
+			cancelCtx, cancel := context.WithCancel(ctx)
+			cancels = append(cancels, cancel)
+			buyTickets := func() {
+				err := vsp.buyTickets(cancelCtx, passphrase, sdiff, expiry)
+				if err != nil {
+					switch {
+					// silence these errors
+					case errors.Is(err, errors.InsufficientBalance):
+					case errors.Is(err, context.Canceled):
+					case errors.Is(err, context.DeadlineExceeded):
+					default:
+						log.Errorf("[%d] Ticket purchasing failed: %v", wal.ID, err)
+					}
+					if errors.Is(err, errors.Passphrase) {
+						fatalMu.Lock()
+						fatal = err
+						fatalMu.Unlock()
+						outerCancel()
+					}
+				}
+			}
+
+			// start separate ticket purchase for as many tickets that can be purchased
+			// each purchase only buy 1 ticket.
+			for i := 0; i < buy; i++ {
+				go buyTickets()
+			}
+		}
+	}
+}
+
+// buyTickets purchases one or more tickets depending on the spendable balance of
+// the selected account and the specified minimum balance to maintain.
+func (vsp *VSP) buyTickets(ctx context.Context, passphrase []byte, sdiff dcrutil.Amount, expiry int32) error {
+	ctx, task := trace.NewTask(ctx, "ticketbuyer.buy")
+	defer task.End()
+
+	wal := vsp.w
+
+	if len(passphrase) > 0 && wal.IsLocked() {
+		err := wal.UnlockWallet(passphrase)
+		if err != nil {
+			return translateError(err)
+		}
+	}
+
+	request := &w.PurchaseTicketsRequest{
+		Count:                1, // prevent combining multi ticket in one transaction. Buy max 1 per txns
+		SourceAccount:        vsp.purchaseAccount,
+		Expiry:               expiry,
+		MinConf:              wal.RequiredConfirmations(),
+		VSPFeeProcess:        vsp.PoolFee,
+		VSPFeePaymentProcess: vsp.ProcessFee,
+	}
+
+	networkBackend, err := wal.Internal().NetworkBackend()
+	if err != nil {
+		return err
+	}
+
+	tix, err := wal.Internal().PurchaseTickets(ctx, networkBackend, request)
+	if tix != nil {
+		for _, hash := range tix.TicketHashes {
+			log.Infof("Purchased ticket %v at stake difficulty %v from wallet: %s", hash, sdiff, wal.Name)
+		}
+	}
+
+	return err
+}
+
+// IsAutoTicketsPurchaseActive returns true if ticket buyer is active
+func (wallet *Wallet) IsAutoTicketsPurchaseActive() bool {
+	wallet.cancelAutoTicketBuyerMux.Lock()
+	defer wallet.cancelAutoTicketBuyerMux.Unlock()
+	return wallet.cancelAutoTicketBuyer != nil
+}
+
+// StopAutoTicketsPurchase stops the active ticket buyer
+func (mw *MultiWallet) StopAutoTicketsPurchase(walletID int) error {
+	wallet := mw.WalletWithID(walletID)
+	if wallet == nil {
+		return errors.New(ErrNotExist)
+	}
+
+	wallet.cancelAutoTicketBuyerMux.Lock()
+	cancelFunc := wallet.cancelAutoTicketBuyer
+	wallet.cancelAutoTicketBuyerMux.Unlock()
+	if cancelFunc == nil {
+		return errors.New(ErrInvalid)
+	}
+
+	wallet.cancelAutoTicketBuyerMux.Lock()
+	wallet.cancelAutoTicketBuyer()
+	wallet.cancelAutoTicketBuyer = nil
+	wallet.cancelAutoTicketBuyerMux.Unlock()
```

Comment:
Simplify. Also, best to keep the lock until you're all done to avoid another process attempting to acquire the lock after you release it midway here.
```suggestion
	wallet.cancelAutoTicketBuyerMux.Lock()
	if wallet.cancelAutoTicketBuyer == nil {
		wallet.cancelAutoTicketBuyerMux.Unlock()	
		return errors.New(ErrInvalid)
	}
	
	wallet.cancelAutoTicketBuyer()
	wallet.cancelAutoTicketBuyer = nil
	wallet.cancelAutoTicketBuyerMux.Unlock()
```